thunder/clang/__init__.py

```diff
@@ -598,7 +598,18 @@ def _basic_indexing(a: TensorLike, /, key) -> TensorLike:
     if key is None or isinstance(key, (Number, NumberProxy, slice, EllipsisType)):
         key = (key,)
 
+    _key = []
+    # keeps track of keys that are not squeeze related
+    new_key = []
     for idx, x in enumerate(key):
+        if not isinstance(x, (NumberProxy, Number)):
+            new_key.append(x)
+        if isinstance(x, (list, TensorLike)):
```

Comment:
My big questions here are:

- does PyTorch's getitem handle those cases correctly? If not, should happen? Should we try to mimic PyTorch's behavior, or mimic NumPy's behavior?
- what combined indexing case or cases does this PR want to support?

See https://github.com/Lightning-AI/lightning-thunder/blob/6dfe7e939a19d1ef5ab259de8709a79f0104fa42/thunder/tests/opinfos.py#L3502 for some existing discrepancies between PyTorch and NumPy.

I think we previously decided (maybe @nikitaved remembers better) that we should implement NumPy's indexing behavior, and we can leave being consistent with PyTorch (or not) for future work.

The NumPy documentation describes two cases of combined indexing here https://numpy.org/doc/stable/user/basics.indexing.html#combining-advanced-and-basic-indexing:

- the advanced indices are separated by a slice, ellipsis, or newaxis
- the advanced indices are all next to each other

Is this PR addressing one or both of those cases?

With basic indexing, should we require that the advanced indexing elements of a key be replaced by ":" before calling the basic indexing operation? That might be more explicit if someone is trying to call it directly, and would avoid this check in basic indexing.

Reply:
To answer  to best of my knowledge, here are my takes:
-  Thunder currently handles both slice and newaxis as basic indices. (see below)
https://github.com/Lightning-AI/lightning-thunder/blob/7d62ae1c7da8fb57e09da3e95ced734e34480400/thunder/clang/__init__.py#L598
https://github.com/Lightning-AI/lightning-thunder/blob/7d62ae1c7da8fb57e09da3e95ced734e34480400/thunder/clang/__init__.py#L649
- Ellipsis are handled in both advanced and basic indexing
- This PR does not necessarily assume that the advanced indices are all next to each other. That is why I had the `transpose` operation. (this allows us to rearrange so that basic indices come first and advanced indices come after; which leads to 'advanced indices are all next to each other')

Reply:
NumPy is the definition of (advanced)indexing (https://numpy.org/doc/stable/user/basics.indexing.html) without any surprises. PyTorch tries to comply with NumPy. If not today, it will at some point. Sticking to NumPy is robust and future-proof.